<a href="https://cognitiveclass.ai"><img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork/images/IDSN-logo.png" width="400"> </a>

# Convolutional Neural Networks with Keras

Estimated time needed **30** mins


In this lab, we will learn how to use the Keras library to build convolutional neural networks. We will also use the popular MNIST dataset and we will compare our results to using a conventional neural network.


## Objectives for this Notebook    
* How to use the Keras library to build convolutional neural networks
* Convolutional neural network with one set of convolutional and pooling layers
* Convolutional neural network with two sets of convolutional and pooling layers



## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
      
1. <a href="#Import-Keras-and-Packages">Import Keras and Packages</a>   
2. <a href="#Convolutional-Neural-Network-with-One-Set-of-Convolutional-and-Pooling-Layers">Convolutional Neural Network with One Set of Convolutional and Pooling Layers</a>  
3. <a href="#Convolutional-Neural-Network-with-Two-Sets-of-Convolutional-and-Pooling-Layers">Convolutional Neural Network with Two Sets of Convolutional and Pooling Layers</a>  

</font>
</div>


### Install the necessary libraries


Let's start by installing the keras libraries and the packages that we would need to build a neural network.


In [ ]:
# All Libraries required for this lab are listed below. The libraries need to be installed on Skills Network Labs. 
# If you run this notebook on a different environment, e.g. your desktop, you may want to install these.
!pip install numpy==2.0.2
!pip install pandas==2.2.2
!pip install tensorflow_cpu==2.18.0
!pip install matplotlib==3.9.2

#### Suppress the tensorflow warning messages
We use the following code to  suppress the warning messages due to use of CPU architechture for tensoflow.
You may want to **comment out** these lines if you are using the GPU architechture


In [3]:
import os
os.environ['TF_ENABLE_ONEDNN_OPTS'] = '0'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'

## Import Keras and Packages


In [4]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Input
from keras.utils import to_categorical

When working with convolutional neural networks in particular, we will need additional packages.


In [5]:
from keras.layers import Conv2D # to add convolutional layers
from keras.layers import MaxPooling2D # to add pooling layers
from keras.layers import Flatten # to flatten data for fully connected layers

## Convolutional Neural Network with One Set of Convolutional and Pooling Layers


In [6]:
# import data
from keras.datasets import mnist

# load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# reshape to be [samples][pixels][width][height]
X_train = X_train.reshape(X_train.shape[0], 28, 28, 1).astype('float32')
X_test = X_test.reshape(X_test.shape[0], 28, 28, 1).astype('float32')

Let's normalize the pixel values to be between 0 and 1


In [7]:
X_train = X_train / 255 # normalize training data
X_test = X_test / 255 # normalize test data

Next, let's convert the target variable into binary categories


In [8]:
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

num_classes = y_test.shape[1] # number of categories

Next, let's define a function that creates our model. Let's start with one set of convolutional and pooling layers.


In [9]:
def convolutional_model():
    
    # create model
    model = Sequential()
    model.add(Input(shape=(28, 28, 1)))
    model.add(Conv2D(16, (5, 5), strides=(1, 1), activation='relu'))
    model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
    
    model.add(Flatten())
    model.add(Dense(100, activation='relu'))
    model.add(Dense(num_classes, activation='softmax'))
    
    # compile model
    model.compile(optimizer='adam', loss='categorical_crossentropy',  metrics=['accuracy'])
    return model

Finally, let's call the function to create the model, and then let's train it and evaluate it.


In [10]:
# build the model
model = convolutional_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=200, verbose=2)

# evaluate the model
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: {} \n Error: {}".format(scores[1], 100-scores[1]*100))

Epoch 1/10
300/300 - 13s - 42ms/step - accuracy: 0.9154 - loss: 0.2957 - val_accuracy: 0.9674 - val_loss: 0.1108
Epoch 2/10
300/300 - 10s - 33ms/step - accuracy: 0.9739 - loss: 0.0900 - val_accuracy: 0.9759 - val_loss: 0.0775
Epoch 3/10
300/300 - 15s - 51ms/step - accuracy: 0.9829 - loss: 0.0584 - val_accuracy: 0.9807 - val_loss: 0.0574
Epoch 4/10
300/300 - 14s - 48ms/step - accuracy: 0.9866 - loss: 0.0442 - val_accuracy: 0.9842 - val_loss: 0.0481
Epoch 5/10
300/300 - 10s - 32ms/step - accuracy: 0.9888 - loss: 0.0367 - val_accuracy: 0.9834 - val_loss: 0.0476
Epoch 6/10
300/300 - 10s - 32ms/step - accuracy: 0.9907 - loss: 0.0301 - val_accuracy: 0.9857 - val_loss: 0.0406
Epoch 7/10
300/300 - 11s - 38ms/step - accuracy: 0.9925 - loss: 0.0251 - val_accuracy: 0.9877 - val_loss: 0.0381
Epoch 8/10
300/300 - 10s - 35ms/step - accuracy: 0.9941 - loss: 0.0196 - val_accuracy: 0.9857 - val_loss: 0.0436
Epoch 9/10
300/300 - 10s - 33ms/step - accuracy: 0.9952 - loss: 0.0165 - val_accuracy: 0.9865 - 

------------------------------------------


## Convolutional Neural Network with Two Sets of Convolutional and Pooling Layers


Let's redefine our convolutional model so that it has two convolutional and pooling layers instead of just one layer of each.


In [11]:
def convolutional_model():
    
    # create model
    model = Sequential()
    model.add(Input(shape=(28, 28, 1)))
    model.add(Conv2D(16, (5, 5), activation='relu'))
    model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
    
    model.add(Conv2D(8, (2, 2), activation='relu'))
    model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))
    
    model.add(Flatten())
    model.add(Dense(100, activation='relu'))
    model.add(Dense(num_classes, activation='softmax'))
    
    # Compile model
    model.compile(optimizer='adam', loss='categorical_crossentropy',  metrics=['accuracy'])
    return model

Now, let's call the function to create our new convolutional neural network, and then let's train it and evaluate it.


In [12]:
# build the model
model = convolutional_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=200, verbose=2)

# evaluate the model
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: {} \n Error: {}".format(scores[1], 100-scores[1]*100))

Epoch 1/10
300/300 - 16s - 54ms/step - accuracy: 0.8770 - loss: 0.4513 - val_accuracy: 0.9595 - val_loss: 0.1333
Epoch 2/10
300/300 - 15s - 48ms/step - accuracy: 0.9661 - loss: 0.1122 - val_accuracy: 0.9717 - val_loss: 0.0853
Epoch 3/10
300/300 - 14s - 47ms/step - accuracy: 0.9748 - loss: 0.0843 - val_accuracy: 0.9796 - val_loss: 0.0663
Epoch 4/10
300/300 - 18s - 60ms/step - accuracy: 0.9790 - loss: 0.0695 - val_accuracy: 0.9802 - val_loss: 0.0609
Epoch 5/10
300/300 - 20s - 67ms/step - accuracy: 0.9816 - loss: 0.0600 - val_accuracy: 0.9808 - val_loss: 0.0542
Epoch 6/10
300/300 - 20s - 67ms/step - accuracy: 0.9831 - loss: 0.0543 - val_accuracy: 0.9842 - val_loss: 0.0484
Epoch 7/10
300/300 - 13s - 45ms/step - accuracy: 0.9859 - loss: 0.0473 - val_accuracy: 0.9846 - val_loss: 0.0437
Epoch 8/10
300/300 - 11s - 37ms/step - accuracy: 0.9870 - loss: 0.0422 - val_accuracy: 0.9830 - val_loss: 0.0477
Epoch 9/10
300/300 - 11s - 35ms/step - accuracy: 0.9875 - loss: 0.0393 - val_accuracy: 0.9871 - 

<h3>Practice Exercise 1</h3>


Let's see how batch size affects the time required and accuracy of the model training. 
For this, you can try to change batch_size to 1024 and check it's effect on accuracy


In [13]:
# build the model
model = convolutional_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=1024, verbose=2)

# evaluate the model
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: {} \n Error: {}".format(scores[1], 100-scores[1]*100))

Epoch 1/10
59/59 - 14s - 239ms/step - accuracy: 0.6662 - loss: 1.2886 - val_accuracy: 0.8803 - val_loss: 0.4299
Epoch 2/10
59/59 - 13s - 223ms/step - accuracy: 0.9053 - loss: 0.3247 - val_accuracy: 0.9329 - val_loss: 0.2292
Epoch 3/10
59/59 - 13s - 227ms/step - accuracy: 0.9407 - loss: 0.2002 - val_accuracy: 0.9562 - val_loss: 0.1507
Epoch 4/10
59/59 - 13s - 227ms/step - accuracy: 0.9571 - loss: 0.1461 - val_accuracy: 0.9652 - val_loss: 0.1192
Epoch 5/10
59/59 - 14s - 230ms/step - accuracy: 0.9656 - loss: 0.1149 - val_accuracy: 0.9712 - val_loss: 0.0975
Epoch 6/10
59/59 - 15s - 250ms/step - accuracy: 0.9711 - loss: 0.0971 - val_accuracy: 0.9748 - val_loss: 0.0840
Epoch 7/10
59/59 - 11s - 193ms/step - accuracy: 0.9753 - loss: 0.0837 - val_accuracy: 0.9766 - val_loss: 0.0750
Epoch 8/10
59/59 - 9s - 159ms/step - accuracy: 0.9776 - loss: 0.0752 - val_accuracy: 0.9776 - val_loss: 0.0705
Epoch 9/10
59/59 - 10s - 169ms/step - accuracy: 0.9800 - loss: 0.0678 - val_accuracy: 0.9797 - val_loss: 

<h3>Practice Exercise 2</h3>


Now, let's see how number of epochs  affect the time required and accuracy of the model training. 
For this, you can keep the batch_size=1024 and epochs=25 and check it's effect on accuracy


In [14]:
# build the model
model = convolutional_model()

# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=25, batch_size=1024, verbose=2)

# evaluate the model
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: {} \n Error: {}".format(scores[1], 100-scores[1]*100))

Epoch 1/25
59/59 - 10s - 172ms/step - accuracy: 0.6571 - loss: 1.2450 - val_accuracy: 0.8860 - val_loss: 0.3867
Epoch 2/25
59/59 - 9s - 152ms/step - accuracy: 0.9122 - loss: 0.2933 - val_accuracy: 0.9414 - val_loss: 0.1960
Epoch 3/25
59/59 - 9s - 148ms/step - accuracy: 0.9476 - loss: 0.1777 - val_accuracy: 0.9579 - val_loss: 0.1347
Epoch 4/25
59/59 - 8s - 142ms/step - accuracy: 0.9620 - loss: 0.1293 - val_accuracy: 0.9692 - val_loss: 0.1036
Epoch 5/25
59/59 - 9s - 152ms/step - accuracy: 0.9692 - loss: 0.1057 - val_accuracy: 0.9727 - val_loss: 0.0915
Epoch 6/25
59/59 - 10s - 163ms/step - accuracy: 0.9725 - loss: 0.0933 - val_accuracy: 0.9766 - val_loss: 0.0785
Epoch 7/25
59/59 - 13s - 217ms/step - accuracy: 0.9757 - loss: 0.0817 - val_accuracy: 0.9791 - val_loss: 0.0701
Epoch 8/25
59/59 - 11s - 184ms/step - accuracy: 0.9779 - loss: 0.0747 - val_accuracy: 0.9805 - val_loss: 0.0633
Epoch 9/25
59/59 - 10s - 168ms/step - accuracy: 0.9802 - loss: 0.0686 - val_accuracy: 0.9815 - val_loss: 0.0

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/). I hope you found this lab interesting and educational. Feel free to contact me if you have any questions!


<!--
## Change Log

|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |
|---|---|---|---|
| 2024-11-20  | 3.0  | Aman  |  Updated the library versions to current |
| 2020-09-21  | 2.0  | Srishti  |  Migrated Lab to Markdown and added to course repo in GitLab |



<hr>

## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>


## <h3 align="center"> &#169; IBM Corporation. All rights reserved. <h3/>

